src/uploads/imageResizer.ts

```diff
@@ -6,128 +6,212 @@ import sharp from 'sharp';
+  name: string,
+  filename: FileSize['filename'] = null,
+  width: FileSize['width'] = null,
+  height: FileSize['height'] = null,
+  filesize: FileSize['filesize'] = null,
+  mimeType: FileSize['mimeType'] = null,
+  sizesToSave: FileToSave[] = [],
+): ResizeResult => ({
+  sizesToSave,
+  sizeData: {
+    [name]: {
+      width,
+      height,
+      filename,
+      filesize,
+      mimeType,
+    },
+  },
+});
+
+/**
+ * Check if the image needs to be resized according to the requested dimensions
+ * and the original image size. If the resize options are provided, the image
+ * will be resized regardless of the requested dimensions, given that the
+ * width or height to be resized is provided.
+ *
+ * @param requestedDimensions - the requested dimensions
+ * @param original - the original image size
+ * @param resizeOptions - the resize options
+ * @returns true if the image needs to be resized, false otherwise
+ */
+const needsResize = (
+  { width, height, withoutEnlargement, withoutReduction }: ImageSize,
+  original: ProbedImageSize,
+): boolean => {
+  // allow enlargement or prevent reduction (our default is to prevent
+  // enlargement and allow reduction)
+  if (withoutEnlargement === false || withoutReduction) {
+    return true;
+  }
+
+  const isWidthOrHeightDefined = !!height || !!width;
+  const hasInsufficientWidth = !!width && width <= original.width;
+  const hasInsufficientHeight = !!height && height <= original.height;
+
+  // If with and height are not defined, it means there is a format conversion
+  // and the image needs to be "resized" (transformed).
+  return !isWidthOrHeightDefined || hasInsufficientWidth || hasInsufficientHeight;
+};
+
+/**
+ * Resize the image and provide the resize buffer. The image will be resized
+ * according to the provided resize config. If no image sizes are requested,
+ * the resolved data will be empty. For every image that dos not need to be
+ * resized, an result object will `null` parameters will be returned.
+ *
+ * @param resizeConfig - the resize config
+ * @returns the result of the resize operation(s)
+ */
 export default async function resizeAndSave({
   req,
   file,
   dimensions,
   staticPath,
   config,
   savedFilename,
-}: Args): Promise<Result> {
+  mimeType,
+}: ResizeArgs): Promise<ResizeResult> {
   const { imageSizes } = config.upload;
-  const sizesToSave: FileToSave[] = [];
-  const sizeData = {};
+
+  // Noting to resize here so return as early as possible
+  if (!imageSizes) return { sizeData: {}, sizesToSave: [] };
 
   const sharpBase = sharp(file.tempFilePath || file.data);
 
-  const promises = imageSizes
-    .map(async (desiredSize) => {
-      if (!needsResize(desiredSize, dimensions)) {
-        sizeData[desiredSize.name] = {
-          url: null,
-          width: null,
-          height: null,
-          filename: null,
-          filesize: null,
-          mimeType: null,
-        };
-        return;
+  const results: ResizeResult[] = await Promise.all(
```

Comment:
Since this does not only do resizing, but also formatting and trimming, I think this could be renamed to something more general instead of `ResizeResult`. Same thing for the function name.

What do you think?